core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -163,21 +188,100 @@ private TableMetadata internalApply() {
     this.base = ops.refresh();
 
     Set<Long> idsToRetain = Sets.newHashSet();
-    List<Long> ancestorIds = SnapshotUtil.ancestorIds(base.currentSnapshot(), base::snapshot);
-    if (minNumSnapshots >= ancestorIds.size()) {
+    // Identify refs that should be removed
+    Set<String> expiredRefs = computeExpiredRefs(base.refs());
+    Set<Long> idsOfRetainedRefs = base.refs().entrySet().stream()
+        .filter(ref -> !expiredRefs.contains(ref.getKey()))
+        .map(refEntry -> refEntry.getValue().snapshotId())
+        .collect(Collectors.toSet());
+
+    for (Long idToRemove : idsToRemove) {
+      Preconditions.checkArgument(!idsOfRetainedRefs.contains(idToRemove),
+          "Cannot expire %s. There is a reference to it", idToRemove);
+    }
+
+    idsToRetain.addAll(idsOfRetainedRefs);
+
+    List<Map.Entry<String, SnapshotRef>> branches = base.refs().entrySet().stream()
+        .filter(refEntry -> refEntry.getValue().isBranch())
+        .collect(Collectors.toList());
+
+    Set<Long> branchSnapshotsToRetain = computeAllBranchSnapshotsToRetain(branches);
+    idsToRetain.addAll(branchSnapshotsToRetain);
+    TableMetadata.Builder updatedMetaBuilder = TableMetadata.buildFrom(base);
+    idsToRemove.addAll(base.snapshots().stream().map(Snapshot::snapshotId)
+        .filter(snapshot -> !idsToRetain.contains(snapshot)).collect(Collectors.toSet()));
+    updatedMetaBuilder.removeSnapshots(idsToRemove);
+    expiredRefs.forEach(updatedMetaBuilder::removeRef);
+    TableMetadata updatedMetadata = updatedMetaBuilder.build();
+    List<Snapshot> updateSnapshots = updatedMetadata.snapshots();
+    List<Snapshot> baseSnapshots = base.snapshots();
+    boolean updated = updateSnapshots.size() != baseSnapshots.size() || expiredRefs.size() > 0;
+    return updated ? updatedMetadata : base;
+  }
+
+  private Set<Long> computeAllBranchSnapshotsToRetain(List<Map.Entry<String, SnapshotRef>> branches) {
+    Set<Long> branchSnapshotsToRetain = Sets.newHashSet();
+    for (Map.Entry<String, SnapshotRef> branchEntry : branches) {
+      String name = branchEntry.getKey();
+      SnapshotRef branch = branchEntry.getValue();
+      Long expireSnapshotsOlderThan;
```

Comment:
I think this should use the same logic that you have for `maxRefAgeMs`:

```
long maxSnapshotAgeMs = branch.maxSnapshotAgeMs() != null ? branch.maxSnapshotAgeMs() : defaultMaxSnapshotAgeMs;
long minSnapshotsToKeep = branch.minSnapshotsToKeep() != null ? branch.minSnapshotsToKeep() : defaultMinSnapshotsToKeep;
```

And those defaults should be from table settings. I don't think we treat `main` differently other than ensuring that we don't age it off.

Reply:
Correct, I commented this earlier, something I need to fix: https://github.com/apache/iceberg/pull/4578/files#r862970455